backends/ref/ceed-ref-operator.c

```diff
@@ -668,6 +668,403 @@ static int CeedOperatorAssembleLinearQFunction_Ref(CeedOperator op,
+  // -- Diagonalize
+  ierr = CeedSimultaneousDiagonalization(ceed, laplace, mass, x, lambda, P1d);
+  CeedChk(ierr);
+  ierr = CeedFree(&work); CeedChk(ierr);
+  ierr = CeedFree(&mass); CeedChk(ierr);
+  ierr = CeedFree(&laplace); CeedChk(ierr);
+  for (CeedInt i=0; i<P1d; i++)
+    for (CeedInt j=0; j<P1d; j++)
+      x2[i+j*P1d] = x[j+i*P1d];
+  ierr = CeedFree(&x); CeedChk(ierr);
+
+  // Assemble QFunction
+  CeedVector assembled;
+  CeedElemRestriction rstr_qf;
+  ierr =  CeedOperatorAssembleLinearQFunction(op, &assembled, &rstr_qf,
+          request); CeedChk(ierr);
+  ierr = CeedElemRestrictionDestroy(&rstr_qf); CeedChk(ierr);
+
+  // Calculate element averages
+  CeedInt nfields = ((interp?1:0) + (grad?dim:0))*((interp?1:0) + (grad?dim:0));
+  CeedScalar *elemavg;
+  const CeedScalar *assembledarray, *qweightsarray;
+  CeedVector qweights;
+  ierr = CeedVectorCreate(ceedparent, nqpts, &qweights); CeedChk(ierr);
+  ierr = CeedBasisApply(basis, 1, CEED_NOTRANSPOSE, CEED_EVAL_WEIGHT,
+                        CEED_VECTOR_NONE, qweights); CeedChk(ierr);
+  ierr = CeedVectorGetArrayRead(assembled, CEED_MEM_HOST, &assembledarray);
+  CeedChk(ierr);
+  ierr = CeedVectorGetArrayRead(qweights, CEED_MEM_HOST, &qweightsarray);
+  CeedChk(ierr);
+  ierr = CeedCalloc(nelem, &elemavg); CeedChk(ierr);
+  for (CeedInt e=0; e<nelem; e++) {
+    CeedInt count = 0;
+    for (CeedInt q=0; q<nqpts; q++)
+      for (CeedInt i=0; i<ncomp*ncomp*nfields; i++)
+        if (fabs(assembledarray[e*nelem*nqpts*ncomp*ncomp*nfields +
+                                                                  i*nqpts + q]) > CEED_EPSILON) {
+          elemavg[e] += assembledarray[e*nelem*nqpts*ncomp*ncomp*nfields +
+                                       i*nqpts + q] / qweightsarray[q];
+          count++;
+        }
+    if (count)
+      elemavg[e] /= count;
+  }
+  ierr = CeedVectorRestoreArrayRead(assembled, &assembledarray); CeedChk(ierr);
+  ierr = CeedVectorDestroy(&assembled); CeedChk(ierr);
+  ierr = CeedVectorRestoreArrayRead(qweights, &qweightsarray); CeedChk(ierr);
+  ierr = CeedVectorDestroy(&qweights); CeedChk(ierr);
+
+  // Build FDM diagonal
+  CeedVector qdata;
+  CeedScalar *qdataarray;
+  ierr = CeedVectorCreate(ceedparent, nelem*ncomp*nnodes, &qdata); CeedChk(ierr);
+  ierr = CeedVectorSetArray(qdata, CEED_MEM_HOST, CEED_COPY_VALUES, NULL);
+  CeedChk(ierr);
+  ierr = CeedVectorGetArray(qdata, CEED_MEM_HOST, &qdataarray); CeedChk(ierr);
+  for (CeedInt e=0; e<nelem; e++)
+    for (CeedInt c=0; c<ncomp; c++)
+      for (CeedInt n=0; n<nnodes; n++) {
+        if (interp)
+          qdataarray[(e*ncomp+c)*nnodes+n] = 1;
+        if (grad)
+          for (CeedInt d=0; d<dim; d++) {
+            CeedInt i = (n / CeedIntPow(P1d, d)) % P1d;
+            qdataarray[(e*ncomp+c)*nnodes+n] += lambda[i];
+          }
+        qdataarray[(e*ncomp+c)*nnodes+n] = 1 / (elemavg[e] *
+                                                qdataarray[(e*ncomp+c)*nnodes+n]);
```

Comment:
What is the performance hot-spot of this PR? I'd be interested in seeing a profile.

Reply:
I'm pretty sure its that loop on lines `797-854`. The function `CeedOperatorAssembleLinearDiagonal` accounts for 80% of the run time.